File: Part8/Go/FunctionConcepts/CompileErrors.go
// Go Function Errors - All errors active for compilation

package main

import "fmt"

// Error: Opening brace on next line (Go requires it on same line)
func printHello()
{
    fmt.Println("Hello")
}

// Error: Missing parameter type
func calculateTotal(size, toppings int, delivery) float64 {
    return 15.99
}

// Error: Return type before function name
float64 func getPrice() {
    return 10.50
}

// Error: Using 'void' instead of no return type
func void sayHi() {
    fmt.Println("Hi")
}

// Error: Using 'none' for no return type
func none display() {
    fmt.Println("Display")
}

// Error: Missing return statement in function with return type
func getNumber() int {
    x := 5
}

// Error: Returning value from function with no return type
func processData() {
    return 42
}

// Error: Return type mismatch
func getValue() int {
    return 3.14
}

// Error: Using 'call' keyword
func myFunction() {
    fmt.Println("Test")
}

func test() {
    call myFunction()
}

// Error: Using semicolon between parameters
func setPizza(size string; count int) {
    fmt.Printf("%s %d\n", size, count)
}

// Error: Parameter type before name
func addNumbers(int a, int b) int {
    return a + b
}

// Error: Default parameters (not supported in Go)
func orderPizza(size string = "medium", toppings int) {
    fmt.Println("Order placed")
}

// Error: Named parameters in call (not supported in Go)
func makePizza(size string, count int) {
    fmt.Printf("%s %d\n", size, count)
}

func placeOrder() {
    makePizza(size = "large", count = 3)
}

// Error: Function overloading (not supported in Go)
func calculate(a int, b int) int {
    return a + b
}

func calculate(a float64, b float64) float64 {
    return a + b
}

// Error: Using 'begin' and 'end' keywords
func compute() begin
    x := 5
end

// Error: Missing comma between parameters
func setValues(count int price float64) {
    fmt.Printf("%d %.2f\n", count, price)
}

// Error: Semicolon after function signature
func doSomething(); {
    fmt.Println("Something")
}

// Error: Using 'function' keyword
function display() {
    fmt.Println("Display")
}

// Error: Using 'procedure' keyword
procedure run() {
    fmt.Println("Running")
}

// Error: Parentheses required even with no parameters (missing them)
func show {
    fmt.Println("Show")
}

// Error: Using 'with parameters' syntax
func execute() {
    fmt.Println("Execute")
}

func caller() {
    call execute with parameters()
}

// Error: Missing parentheses in function call
func greet() {
    fmt.Println("Greet")
}

func callGreet() {
    greet
}

// Error: Wrong number of arguments
func add(a int, b int) int {
    return a + b
}

func callAdd() {
    result := add(5)
}

// Error: Using := for parameter default (trying default params wrong way)
func createOrder(size string := "small") {
    fmt.Println(size)
}

// Error: Type in function call argument
func getTotal(price float64, tax float64) float64 {
    return price + tax
}

func purchase() {
    total := getTotal(float64 10.0, float64 2.0)
}

// Error: Multiple return values without parentheses
func divmod(a int, b int) int, int {
    return a / b, a % b
}

// Error: Return statement outside function
var globalValue = 10
return globalValue

// Error: Using 'def' keyword
def printMessage() {
    fmt.Println("Message")
}

// Error: Nested function definition (not standard Go)
func outer() {
    func inner() {
        fmt.Println("Inner")
    }
}

func main() {
    fmt.Println("Main")
}

File: Part8/C++/FunctionConcepts/CompileErrors.cpp
// C++ Function Errors - All errors active for compilation

#include <iostream>
#include <string>
using namespace std;

// Error: Missing return type
printHello() {
    cout << "Hello" << endl;
}

// Error: Missing semicolon after function definition
void sayHi() {
    cout << "Hi" << endl;
}

void greet() {
    cout << "Greetings" << endl;
}

// Error: Function defined inside another function (not allowed in standard C++)
void outerFunction() {
    void innerFunction() {
        cout << "Inner" << endl;
    }
}

// Error: Missing parameter type
float calculateTotal(size, int toppings) {
    return 15.99;
}

// Error: Return type doesn't match returned value
int getValue() {
    return 3.14;
}

// Error: Missing return statement in non-void function
int getNumber() {
    int x = 5;
}

// Error: Returning value from void function
void processData() {
    return 42;
}

// Error: Using 'call' keyword before function name
void myFunction() {
    cout << "Test" << endl;
}

void test() {
    call myFunction();
}

// Error: Function overloading with same signature (only return type differs)
int calculate(int a, int b) {
    return a + b;
}

double calculate(int a, int b) {
    return a + b;
}

// Error: Missing parentheses in function call
void display() {
    cout << "Display" << endl;
}

void run() {
    display;
}

// Error: Default parameter not at end
void orderPizza(string size = "medium", int toppings, bool delivery = false) {
    cout << "Order placed" << endl;
}

// Error: Redefining default parameter in implementation
void makeOrder(string size, int toppings = 2);

void makeOrder(string size, int toppings = 3) {
    cout << "Making order" << endl;
}

// Error: Missing semicolon after function prototype
void prototype()

// Error: Parameter name used in function call (C++ doesn't support named parameters like this)
void setPizza(string size, int count) {
    cout << size << " " << count << endl;
}

void order() {
    setPizza(size = "large", count = 3);
}

// Error: Using 'begin' and 'end' keywords
void calculate() begin
    int x = 5;
end

// Error: Using 'none' or 'procedure' instead of void
none printMessage() {
    cout << "Message" << endl;
}

procedure doSomething() {
    cout << "Something" << endl;
}

// Error: Missing comma between parameters
int addNumbers(int a int b) {
    return a + b;
}

// Error: Using semicolon instead of comma between parameters
float getPrice(float base; float tax) {
    return base + tax;
}

// Error: Semicolon after function header before opening brace
void processValue(); {
    int x = 10;
}

// Error: Parameter type comes after parameter name
void setValue(count int, price float) {
    cout << count << " " << price << endl;
}

// Error: Using 'default' keyword for default parameters
void createPizza(string size default "small") {
    cout << size << endl;
}

// Error: Using 'with parameters' syntax
void myFunc() {
    cout << "Function" << endl;
}

void caller() {
    call myFunc with parameters();
}

// Error: Missing parameter name in function definition
int multiply(int, int b) {
    return 5 * b;
}

// Error: Wrong number of arguments
int add(int a, int b) {
    return a + b;
}

void callAdd() {
    int result = add(5);
}

// Error: Using 'function' keyword
function void display() {
    cout << "Display" << endl;
}

// Error: Return statement outside function
int globalValue = 10;
return globalValue;

// Error: Type specified in function call arguments
float computeTotal(float price, float tax) {
    return price + tax;
}

void purchase() {
    float total = computeTotal(float 10.0, float 2.0);
}

int main() {
    cout << "Main" << endl;
    return 0;
}
